pkg/kubelet/pod_workers.go

```diff
@@ -734,10 +750,80 @@ func calculateEffectiveGracePeriod(status *podSyncStatus, pod *v1.Pod, options *
 	return gracePeriod, status.gracePeriod != 0 && status.gracePeriod != gracePeriod
 }
 
+// allowPodStart tries to start the pod and returns true if allowed, otherwise
+// it requeues the pod and returns false.
+func (p *podWorkers) allowPodStart(pod *v1.Pod) bool {
+	if !kubetypes.IsStaticPod(pod) {
+		// TBD: Do we want to allow non-static pods with the same full name?
+		// Note that it may disable the force deletion of pods.
+		return true
+	}
+	p.podLock.Lock()
+	defer p.podLock.Unlock()
+	status, ok := p.podSyncStatuses[pod.UID]
+	if !ok {
+		klog.ErrorS(nil, "Failed to get a valid podSyncStatuses", "pod", klog.KObj(pod), "podUID", pod.UID)
+		return false
+	}
+	if !p.isStaticPodAllowedToStart(status.fullname, pod.UID) {
+		p.workQueue.Enqueue(pod.UID, wait.Jitter(p.backOffPeriod, workerBackOffPeriodJitterFactor))
+		status.working = false
+		return false
+	}
+	p.allowNextStaticPod(status.fullname, pod.UID)
+	return true
+}
+
+// isStaticPodAllowedToStart returns true if there is no other started static
+// pod with the same fullname and uid matches that of the first static pod
+// waiting to start.
+func (p *podWorkers) isStaticPodAllowedToStart(fullname string, uid types.UID) bool {
+	startedUID, started := p.startedStaticPodsByFullname[fullname]
+	if started {
+		return startedUID == uid
+	}
+
+	waitingPods := p.waitingToStartStaticPodsByFullname[fullname]
+	if len(waitingPods) == 0 {
+		return true
+	}
+
+	return uid == waitingPods[0]
```

Comment:
I realized after reviewing this more that we need to integrate the cleanup logic into the ability to check the first start.  Because waiting cleanup happens on startup check vs task cleanup, we can collapse most of allowNextStaticPod into here and potentially remove that method altogether.

Roughly, when we check waitingPods here we want to find the first UID that exists, trim anything we skipped past.  I think at that point the bulk of maintenance that is currently in `allowNextStaticPod` can be moved into here.

Roughly:

```
waitingPods := p.waitingToStartStaticPodsByFullname[fullname]
if len(waitingPods) == 0 {
  p.startedStaticPodsByFullname[fullname] = uid
  return true
}
for i, waitingUID := range waitingPods {
  if _, ok := p.podSyncStatuses[waitingUID]; !ok {
    continue
  }
  p.waitingToStartStaticPodsByFullname[fullname] = waitingPods[i+1:]
  if waitingUID == uid {
    p.startedStaticPodsByFullname[fullname] = uid
    return true
  }
  return false
}
delete(p.waitingToStartStaticPodsByFullname, fullname)
p.startedStaticPodsByFullname[fullname] = uid
return true
```

This should guarantee that we trim any pods that are no longer waiting.  Does this make sense?

Reply:
seems fine. Done.

removed verifying the cleanup in `allowNextPodToStart` part as we don't care tail, is this ok?